Applications/Spire/Source/Ui/KeySequenceInputField.cpp

```diff
@@ -0,0 +1,271 @@
+
+namespace {
+  const auto& TEXT_PADDING() {
+    static auto padding = scale_width(4);
+    return padding;
+  }
+
+  auto get_key_color(Qt::Key key, bool is_enabled) {
+    switch(key) {
+      case Qt::Key_Alt:
+        if(is_enabled) {
+          return QColor("#C6A600");
+        }
+        return QColor("#E3D68F");
+      case Qt::Key_Control:
+        if(is_enabled) {
+          return QColor("#00B395");
+        }
+        return QColor("#A6E4DA");
+      case Qt::Key_Shift:
+        if(is_enabled) {
+          return QColor("#4495FF");
+        }
+        return QColor("#B0D2FF");
+    }
+    if(is_enabled) {
+      return QColor("#684BC7");
+    }
+    return QColor("#C0B3E9");
+  }
+
+  auto get_key_text(Qt::Key key) {
+    switch(key) {
+      case Qt::Key_Shift:
+        return QObject::tr("Shift");
+      case Qt::Key_Alt:
+        return QObject::tr("Alt");
+      case Qt::Key_Control:
+        return QObject::tr("Ctrl");
+      case Qt::Key_unknown:
+        return QObject::tr("NaK");
+      default:
+        return QKeySequence(key).toString();
+    }
+  }
+
+  auto make_key_sequence(const std::vector<Qt::Key>& keys) {
+    switch(keys.size()) {
+      case 1:
+        return QKeySequence(keys[0]);
+      case 2:
+        return QKeySequence(keys[0], keys[1]);
+      case 3:
+        return QKeySequence(keys[0], keys[1], keys[2]);
+      case 4:
+        return QKeySequence(keys[0], keys[1], keys[2], keys[3]);
+      default:
+        return QKeySequence();
+    }
+  }
+
+
+  QKeySequence sort_key_sequence(const QKeySequence& sequence) {
+    auto modifier_list = std::vector<Qt::Key>();
+    auto function_list = std::vector<Qt::Key>();
+    for(auto i = 0; i < sequence.count(); ++i) {
+      auto key = Qt::Key(sequence[i]);
+      if(key == Qt::Key_Control || key == Qt::Key_Alt ||
+          key == Qt::Key_Shift) {
+        modifier_list.push_back(key);
+      } else {
+        function_list.push_back(key);
+      }
+    }
+    if(modifier_list.empty()) {
+      return make_key_sequence(function_list);
+    } else if(modifier_list.size() > 1) {
+      std::sort(modifier_list.begin(), modifier_list.end(),
+        [] (auto key1, auto key2) {
+          if(key1 == Qt::Key_Control) {
+            return true;
+          } else if(key2 == Qt::Key_Control) {
+            return false;
+          }
+          return key1 == Qt::Key_Alt;
+        });
+    }
+    modifier_list.insert(modifier_list.end(), function_list.begin(),
+      function_list.end());
+    return make_key_sequence(modifier_list);
+  }
+}
+
+KeySequenceInputField::KeySequenceInputField(
+    std::shared_ptr<KeySequenceValidationModel> model, QWidget* parent)
+    : QLineEdit(parent),
+      m_model(std::move(model)),
+      m_state(State::DEFAULT),
+      m_font("Roboto"),
+      m_last_pressed_key(Qt::Key_unknown),
+      m_is_last_key_event_release(false) {
+  setReadOnly(true);
+  setAttribute(Qt::WA_Hover);
+  m_font.setPixelSize(scale_height(12));
```

Comment:
Why is ```m_font``` a member variable?

Reply:
Should I just construct it in the paintEvent? I remember a long while ago you said that constructing objects was cheap, but is it really more efficient to construct it and call its' setters every repaint versus the memory usage when storing it as a member?